papaya-core/src/main/java/pl/edu/agh/papaya/service/sprint/SprintService.java

```diff
@@ -94,27 +95,58 @@ Sprint createSprint(Sprint sprint) {
         return sprintRepository.findFirstByProjectIdOrderByDurationPeriodStartDesc(projectId);
     }
 
-    public Sprint closeSprint(Sprint sprint, Duration timeBurned, Duration timePlanned, LocalDateTime dateClosed) {
-        sprint.setTimeBurned(timeBurned);
-        sprint.setTimePlanned(timePlanned);
+    public Sprint closeSprint(Sprint sprint, Duration timeBurned, Duration finalTimePlanned, LocalDateTime dateClosed) {
         sprint.setDateClosed(dateClosed);
+        updateSprintStats(sprint, finalTimePlanned, timeBurned);
 
         sprintRepository.save(sprint);
 
         return sprint;
     }
 
-    public Optional<Sprint> getById(Long id) {
-        return sprintRepository.findById(id);
+    public void updateSprintStats(Sprint sprint, Duration finalTimePlanned, Duration timeBurned) {
+        SprintStats stats = sprint.getStats();
+
+        stats.setFinalTimePlanned(finalTimePlanned);
+        stats.setTimeBurned(timeBurned);
+
+        Duration totalDeclaredTime = getTotalDeclaredTimeBySprint(sprint);
+        stats.updateCoefficient(totalDeclaredTime);
+
+        stats.setAverageCoefficientCache(computeAvarageSprintCoefficient(sprint));
+    }
+
+    private Double computeAvarageSprintCoefficient(Sprint sprint) {
+        // date after start of this sprint because this one also has to be included
+        Double coefficient = sprintRepository.findAverageSprintCoefficientUpToDate(sprint.getDurationPeriod().getEnd());
+        if (coefficient == null) {
+            coefficient = getDefaultSprintCoefficient(sprint);
+        }
+
+        return coefficient;
     }
 
-    public Duration getTotalAvailableTimeBySprintId(Long sprintId) {
-        return availabilityService.getBySprintId(sprintId)
+    private double getDefaultSprintCoefficient(Sprint sprint) {
+        return sprint.getProject().getInitialCoefficient();
+    }
+
+    public Duration getTotalDeclaredTimeBySprint(Sprint sprint) {
+        return availabilityService.getBySprintId(sprint.getId())
                 .stream()
                 .map(Availability::getTimeAvailable)
                 .reduce(Duration.ZERO, Duration::plus);
     }
 
+    public Optional<Sprint> getById(Long id) {
+        return sprintRepository.findById(id);
+    }
+
+    public double getPrevSprintAverageCoefficient(Sprint sprint) {
+        Optional<Sprint> prevSprint = sprintRepository.findPreceding(sprint);
+        return prevSprint.map(s -> s.getStats().getAverageCoefficientCache())
```

Comment:
To bym zwinęła do jednej linijki i zmieniłabym `s` w lambdzie na `prevSprint`

(Subiektywne odczucie: `prev` -> `previous`, ale to kompletnie bez znaczenia)

Reply:
> zmieniłabym s w lambdzie na prevSprint

nie da rady bo kolizja nazw

Reply:
spoko